applications/FluidDynamicsApplication/custom_processes/surface_smoothing_process.h

```diff
@@ -0,0 +1,259 @@
+#include "modeler/connectivity_preserve_modeler.h"
+#include "solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h"
+#include "solving_strategies/schemes/residualbased_incrementalupdate_static_scheme.h"
+#include "solving_strategies/strategies/residualbased_linear_strategy.h"
+#include "utilities/variable_utils.h"
+#include "utilities/element_size_calculator.h"
+#include "includes/global_pointer_variables.h"
+
+// Application includes
+#include "fluid_dynamics_application_variables.h"
+#include "custom_elements/surface_smoothing_element.h"
+
+namespace Kratos
+{
+///@addtogroup FluidDynamicsApplication
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Utility for surface smoothing
+/// Based on Tornberg, Anna-Karin, and Björn Engquist. "A finite element based level-set method
+/// for multiphase flow applications." Computing and Visualization in Science 3, no. 1-2 (2000): 93-101.
+/// The algorithm is improved by imposing a boundary condition and a correction step.
+class SurfaceSmoothingProcess : public Process
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of SurfaceSmoothingProcess
+    KRATOS_CLASS_POINTER_DEFINITION(SurfaceSmoothingProcess);
+
+    typedef UblasSpace<double, CompressedMatrix, Vector> TSparseSpace;
+    typedef UblasSpace<double, Matrix, Vector> TDenseSpace;
+    typedef LinearSolver<TSparseSpace, TDenseSpace > TLinearSolver;
+    typedef Scheme< TSparseSpace,  TDenseSpace > SchemeType;
+    typedef typename SchemeType::Pointer SchemePointerType;
+    typedef typename BuilderAndSolver<TSparseSpace,TDenseSpace,TLinearSolver>::Pointer BuilderSolverPointerType;
+    typedef SolvingStrategy< TSparseSpace, TDenseSpace, TLinearSolver > SolvingStrategyType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Constructor.
+    SurfaceSmoothingProcess(
+        ModelPart& rModelPart,
+        TLinearSolver::Pointer);
+
+    /// Destructor.
+    ~SurfaceSmoothingProcess() override
+    {
+        Model& current_model = mrModelPart.GetModel();
+        if(current_model.HasModelPart( mAuxModelPartName ))
+            current_model.DeleteModelPart( mAuxModelPartName );
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    void Execute() override;
+
+    void ExecuteInitialize() override;
+
+    void ExecuteBeforeSolutionLoop() override;
+
+    void ExecuteInitializeSolutionStep() override;
+
+    void ExecuteFinalizeSolutionStep() override;
+
+    virtual void Clear()
+    {
+        Model& r_model = mrModelPart.GetModel();
+        ModelPart& r_smoothing_model_part = r_model.GetModelPart( mAuxModelPartName );
+        r_smoothing_model_part.Nodes().clear();
+        r_smoothing_model_part.Conditions().clear();
+        r_smoothing_model_part.Elements().clear();
+        mp_solving_strategy->Clear();
```

Comment:
Implementations in the source file please.

Reply:
Done.